pinot-controller/src/main/java/org/apache/pinot/controller/api/resources/PinotTableRestletResource.java

```diff
@@ -909,6 +912,51 @@ private void checkHybridTableConfig(String rawTableName, TableConfig tableConfig
     }
   }
 
+  /*
+   * Validates that we have sufficient number of servers in the specified tenants to be able to serve the requested
+   * replication in the table config
+   */
+  private void ensureCanHostAllReplicas(TableConfig tableConfig) {
+    int replication = tableConfig.getReplication();
+    TagOverrideConfig tagOverrideConfig = tableConfig.getTenantConfig().getTagOverrideConfig();
+
+    if (tagOverrideConfig != null && tagOverrideConfig.getRealtimeConsuming() != null
+        && tagOverrideConfig.getRealtimeCompleted() != null) {
+      validateReplicationWithTags(tagOverrideConfig.getRealtimeConsuming(), tagOverrideConfig.getRealtimeCompleted(),
+          replication);
+    } else {
+      String serverTenant = tableConfig.getTenantConfig().getServer();
+      Set<String> instancesInServerTenant = _pinotHelixResourceManager.getAllInstancesForServerTenant(serverTenant);
+      if (instancesInServerTenant.size() < replication) {
+        throw new IllegalStateException(
+            String.format("Insufficient servers (tenant: %s, count: %s) to accommodate the requested replication of %s",
+                serverTenant, instancesInServerTenant.size(), replication));
+      }
+    }
+  }
+
+  private void validateReplicationWithTags(String consumingTag, String completedTag, int replication) {
+    int consumingServersCnt = _pinotHelixResourceManager.getAllInstancesWithTag(consumingTag).size();
+    int completedServersCnt = _pinotHelixResourceManager.getAllInstancesWithTag(completedTag).size();
+
+    if (consumingServersCnt < replication || completedServersCnt < replication) {
+      if (consumingServersCnt < replication && completedServersCnt < replication) {
```

Comment:
OOC, what's the Consuming tag and Completed tag mean and how do they get assigned to servers?